src/kcas.ml

```diff
@@ -194,35 +238,43 @@ module Op = struct
   type t = CAS : 'a Loc.t * 'a * 'a -> t
 
   let make_cas loc before after = CAS (loc, before, after) [@@inline]
+  let make_cmp loc expected = CAS (loc, expected, expected) [@@inline]
 
   let is_on_loc op loc =
     match op with CAS (loc', _, _) -> Obj.magic loc' == loc
     [@@inline]
 
   let atomic = function
-    | CAS (loc, before, after) -> Loc.compare_and_set loc before after
+    | CAS (loc, before, after) ->
+        if before == after then Loc.get loc == before
+        else Loc.compare_and_set loc before after
 
-  let atomically = function
+  let atomically ?(mode = Mode.lock_free) = function
     | [] -> true
     | [ op ] -> atomic op
     | first :: rest ->
-        let casn = Atomic.make `After in
-        let insert cass = function
-          | CAS (loc, before, after) -> insert cass loc { before; after; casn }
+        let casn = Atomic.make (mode :> status) in
+        let rec run cass = function
+          | [] -> determine_for_owner casn cass
+          | CAS (loc, before, after) :: rest ->
+              if before == after && is_obstruction_free casn then
+                let state = Atomic.get loc.state in
+                before == eval state && run (insert cass loc state) rest
+              else run (insert cass loc { before; after; casn }) rest
         in
-        let cass =
-          match first with
-          | CAS (loc, before, after) ->
-              CASN (loc, { before; after; casn }, NIL, NIL)
-        in
-        let cass = List.fold_left insert cass rest in
-        determine_for_owner casn cass
+        let (CAS (loc, before, after)) = first in
+        if before == after && is_obstruction_free casn then
+          let state = Atomic.get loc.state in
+          before == eval state && run (CASN (loc, state, NIL, NIL)) rest
+        else run (CASN (loc, { before; after; casn }, NIL, NIL)) rest
```

Comment:
Can the duplication above (250-263 and 266-269) be avoided?

Reply:
Hmm... The duplication here is a bit unfortunate, but it is difficult to eliminate without adding overhead.  Recursion tends to limit compiler optimizations and there is also nested conditionals here (`if` and also `&& run`).  It is basically an explicit manual unrolling of the logic for the `first` logical CAS and it does give measurable overhead reduction.  It also plays together with the optimization in `insert` that explicitly checks for the left/right leaning trees.  These optimizations (unroll the loop once and special case for presorted) were already introduced in the previous implementation.

BTW, unrelated to this.  I realized that `Tx.get` and `Tx.get_as` can be expressed in terms of `Tx.update_as` and in those cases it doesn't seem to be more expensive, which is probably partly due to those not being recursive functions, so the compiler will be able to optimize them after inlining.  I just removed the duplication and now `Tx.get` and `Tx.get_as` are implemented in terms of the internal `update_as` (all the location accesses now are).